controllers/cruisecontroltask_controller.go

```diff
@@ -62,448 +59,412 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
+	Err       string
+	Operation CruiseControlOperation
 }
 
-// getCorrectRequiredCCState returns the correct Required CC state based on that we upscale or downscale
-func (r *CruiseControlTaskReconciler) getCorrectRequiredCCState(ccState kafkav1beta1.CruiseControlState) (kafkav1beta1.CruiseControlState, error) {
-	if ccState.IsDownscale() {
-		return kafkav1beta1.GracefulDownscaleRequired, nil
-	} else if ccState.IsUpscale() {
-		return kafkav1beta1.GracefulUpscaleRequired, nil
+// IsDone returns true if the task is considered finished.
+func (t *CruiseControlTask) IsDone() bool {
+	switch t.Operation {
+	case OperationAddBroker, OperationRemoveBroker:
+		return !t.BrokerState.IsActive()
+	case OperationRebalanceDisks:
+		return !t.VolumeState.IsActive()
 	}
-
-	return ccState, errors.NewWithDetails("could not determine if cruise control state is upscale or downscale", "ccState", ccState)
+	return false
 }
 
-//TODO merge with checkCCTaskState into one func (hi-im-aren)
-func (r *CruiseControlTaskReconciler) checkVolumeCCTaskState(kafkaCluster *kafkav1beta1.KafkaCluster, brokersVolumesState map[string]map[string]kafkav1beta1.VolumeState, log logr.Logger) error {
-	if len(brokersVolumesState) == 0 {
-		return nil
+// Apply takes a kafkav1beta1.KafkaCluster instance and updates its Status field to reflect the state of the task.
+func (t *CruiseControlTask) Apply(instance *kafkav1beta1.KafkaCluster) {
+	if instance == nil {
+		return
 	}
 
-	var ccTaskId string
-	for _, brokerVolumesState := range brokersVolumesState {
-		for _, volumeState := range brokerVolumesState {
-			if ccTaskId == "" {
-				ccTaskId = volumeState.CruiseControlTaskId
-			} else if ccTaskId != volumeState.CruiseControlTaskId {
-				return errors.New("multiple rebalance disk CC task ids found")
+	switch t.Operation {
+	case OperationAddBroker, OperationRemoveBroker:
+		if state, ok := instance.Status.BrokersState[t.BrokerID]; ok {
+			state.GracefulActionState.CruiseControlState = t.BrokerState
+			state.GracefulActionState.CruiseControlTaskId = t.TaskID
+			state.GracefulActionState.TaskStarted = t.StartedAt
+			state.GracefulActionState.ErrorMessage = t.Err
+			instance.Status.BrokersState[t.BrokerID] = state
+		}
+	case OperationRebalanceDisks:
+		if state, ok := instance.Status.BrokersState[t.BrokerID]; ok {
+			if volState, ok := state.GracefulActionState.VolumeStates[t.Volume]; ok {
+				volState.CruiseControlVolumeState = t.VolumeState
+				volState.CruiseControlTaskId = t.TaskID
+				volState.TaskStarted = t.StartedAt
+				volState.ErrorMessage = t.Err
+				instance.Status.BrokersState[t.BrokerID].GracefulActionState.VolumeStates[t.Volume] = volState
 			}
 		}
 	}
+}
 
-	if ccTaskId == "" {
-		return errors.New("no CC task id provided to be checked")
+// FromResult takes a scale.Result instance returned by scale.CruiseControlScaler and updates its own state accordingly.
+func (t *CruiseControlTask) FromResult(result *scale.Result) {
+	if result == nil {
```

Comment:
```
if t == nil || result == nil {
  return
```
just in case this method is invoked on a nil reference

Reply:
Nice catch. Fixed it.